clang/test/CodeGen/X86/bfloat16-convert-half.c

```diff
@@ -0,0 +1,165 @@
+// CHECK-NEXT:    [[TMP0:%.*]] = bitcast <2 x bfloat> [[IN1]] to <2 x half>
+// CHECK-NEXT:    store <2 x half> [[TMP0]], ptr [[RETVAL]], align 4
+// CHECK-NEXT:    [[TMP1:%.*]] = load i32, ptr [[RETVAL]], align 4
+// CHECK-NEXT:    ret i32 [[TMP1]]
+//
+half2 test_cast_from_bfloat2_to_half2(bfloat2 in) {
+  return (half2)in;
+}
+
+
+// CHECK-LABEL: define dso_local double @test_cast_from_bfloat4_to_half4(
+// CHECK-SAME: double noundef [[IN_COERCE:%.*]]) #[[ATTR0]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[RETVAL:%.*]] = alloca <4 x half>, align 8
+// CHECK-NEXT:    [[IN:%.*]] = alloca <4 x bfloat>, align 8
+// CHECK-NEXT:    store double [[IN_COERCE]], ptr [[IN]], align 8
+// CHECK-NEXT:    [[IN1:%.*]] = load <4 x bfloat>, ptr [[IN]], align 8
+// CHECK-NEXT:    [[TMP0:%.*]] = bitcast <4 x bfloat> [[IN1]] to <4 x half>
+// CHECK-NEXT:    store <4 x half> [[TMP0]], ptr [[RETVAL]], align 8
+// CHECK-NEXT:    [[TMP1:%.*]] = load double, ptr [[RETVAL]], align 8
+// CHECK-NEXT:    ret double [[TMP1]]
+//
+half4 test_cast_from_bfloat4_to_half4(bfloat4 in) {
+  return (half4)in;
+}
+
+
+// CHECK-LABEL: define dso_local i32 @test_convertvector_from_half2_to_bfloat2(
+// CHECK-SAME: i32 noundef [[IN_COERCE:%.*]]) #[[ATTR0]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[RETVAL:%.*]] = alloca <2 x bfloat>, align 4
+// CHECK-NEXT:    [[IN:%.*]] = alloca <2 x half>, align 4
+// CHECK-NEXT:    store i32 [[IN_COERCE]], ptr [[IN]], align 4
+// CHECK-NEXT:    [[IN1:%.*]] = load <2 x half>, ptr [[IN]], align 4
+// CHECK-NEXT:    [[FPEXT:%.*]] = fpext <2 x half> [[IN1]] to <2 x float>
+// CHECK-NEXT:    [[FPTRUNC:%.*]] = fptrunc <2 x float> [[FPEXT]] to <2 x bfloat>
+// CHECK-NEXT:    store <2 x bfloat> [[FPTRUNC]], ptr [[RETVAL]], align 4
+// CHECK-NEXT:    [[TMP0:%.*]] = load i32, ptr [[RETVAL]], align 4
+// CHECK-NEXT:    ret i32 [[TMP0]]
+//
+bfloat2 test_convertvector_from_half2_to_bfloat2(half2 in) {
+  return __builtin_convertvector(in, bfloat2);
+}
+
+// CHECK-LABEL: define dso_local i32 @test_convertvector_from_bfloat2_to_half2(
+// CHECK-SAME: i32 noundef [[IN_COERCE:%.*]]) #[[ATTR0]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[RETVAL:%.*]] = alloca <2 x half>, align 4
+// CHECK-NEXT:    [[IN:%.*]] = alloca <2 x bfloat>, align 4
+// CHECK-NEXT:    store i32 [[IN_COERCE]], ptr [[IN]], align 4
+// CHECK-NEXT:    [[IN1:%.*]] = load <2 x bfloat>, ptr [[IN]], align 4
+// CHECK-NEXT:    [[FPEXT:%.*]] = fpext <2 x bfloat> [[IN1]] to <2 x float>
+// CHECK-NEXT:    [[FPTRUNC:%.*]] = fptrunc <2 x float> [[FPEXT]] to <2 x half>
+// CHECK-NEXT:    store <2 x half> [[FPTRUNC]], ptr [[RETVAL]], align 4
+// CHECK-NEXT:    [[TMP0:%.*]] = load i32, ptr [[RETVAL]], align 4
+// CHECK-NEXT:    ret i32 [[TMP0]]
+//
+half2 test_convertvector_from_bfloat2_to_half2(bfloat2 in) {
+  return __builtin_convertvector(in, half2);
+}
+
+// CHECK-LABEL: define dso_local double @test_convertvector_from_half4_to_bfloat4(
```

Comment:
Is the double here some weird ABI thing or intentional?

Reply:
It follows X86's ABI I think.